All calls to ML pipeline scripts use the filepath on my personal machine; these need to be changed if you're re-running the analysis on a different machine

# Approach 1: RGB data only from Oct 7

In [1]:
import pandas as pd

In [3]:
data = pd.read_csv('module2_data/CleanData_InputML.txt',sep='\t')

In [4]:
data.head()

PlotID  Accession Genotype AltGenotype  EarLeafArea  TotalLeafAreaPlot  \
0  EL19_5166  Ames10254     GE54        GE54      630.330          15758.250   
1  EL19_6459  Ames10254     GE54        GE54      557.540          15053.580   
2  EL19_5715  Ames10256    GE129       GE129      470.760          15064.320   
3  EL19_6527  Ames10256    GE129       GE129      442.375          13713.625   
4  EL19_5847  Ames12726     NC13        NC13      524.775          17317.575   

   Min_RGB_10_7_19_R  Min_RGB_10_7_19_G  Min_RGB_10_7_19_B  \
0                  0                  0                  0   
1                  0                  0                  0   
2                 19                 17                 17   
3                 27                 28                 25   
4                 35                 27                 24   

   1stQu_RGB_10_7_19_R  ...  Mean_RGB_10_7_19_B  3rdQu_RGB_10_7_19_R  \
0                64.75  ...           87.992647               138.00   
1                70.00  ...           91.296296               134.50   
2                74.25  ...           96.047619               175.75   
3                99.75  ...          107.442857               177.00   
4                75.25  ...           92.301587               152.00   

   3rdQu_RGB_10_7_19_G  3rdQu_RGB_10_7_19_B  Max_RGB_10_7_19_R  \
0               129.25               113.50                254   
1               127.00               117.50                251   
2               180.00               131.00                250   
3               182.00               137.50                240   
4               144.00               124.75                235   

   Max_RGB_10_7_19_G  Max_RGB_10_7_19_B  sd_RGB_10_7_19_R  sd_RGB_10_7_19_G  \
0                254                237         57.046799         57.085175   
1                255                240         53.171706         54.352414   
2                251                207         62.698556         64.904915   
3                246                210         49.951415         54.026298   
4                232                207         49.409935         49.958817   

   sd_RGB_10_7_19_B  
0         50.170302  
1         48.476502  
2         50.130367  
3         41.406264  
4         42.085108  

[5 rows x 27 columns]

In [5]:
# use PlotID as ID, TotalLeafAreaPlot as label, drop all other non-data columns 
data.drop(labels = ['Accession','Genotype','AltGenotype','EarLeafArea'],axis = 1,inplace=True)
data.head()

PlotID  TotalLeafAreaPlot  Min_RGB_10_7_19_R  Min_RGB_10_7_19_G  \
0  EL19_5166          15758.250                  0                  0   
1  EL19_6459          15053.580                  0                  0   
2  EL19_5715          15064.320                 19                 17   
3  EL19_6527          13713.625                 27                 28   
4  EL19_5847          17317.575                 35                 27   

   Min_RGB_10_7_19_B  1stQu_RGB_10_7_19_R  1stQu_RGB_10_7_19_G  \
0                  0                64.75                56.75   
1                  0                70.00                61.00   
2                 17                74.25                72.50   
3                 25                99.75                99.25   
4                 24                75.25                66.25   

   1stQu_RGB_10_7_19_B  Median_RGB_10_7_19_R  Median_RGB_10_7_19_G  ...  \
0                52.00                  93.0                  85.0  ...   
1                54.50                  96.0                  91.0  ...   
2                55.25                 120.5                 122.0  ...   
3                75.75                 131.0                 127.5  ...   
4                60.25                 100.5                  92.0  ...   

   Mean_RGB_10_7_19_B  3rdQu_RGB_10_7_19_R  3rdQu_RGB_10_7_19_G  \
0           87.992647               138.00               129.25   
1           91.296296               134.50               127.00   
2           96.047619               175.75               180.00   
3          107.442857               177.00               182.00   
4           92.301587               152.00               144.00   

   3rdQu_RGB_10_7_19_B  Max_RGB_10_7_19_R  Max_RGB_10_7_19_G  \
0               113.50                254                254   
1               117.50                251                255   
2               131.00                250                251   
3               137.50                240                246   
4               124.75                235                232   

   Max_RGB_10_7_19_B  sd_RGB_10_7_19_R  sd_RGB_10_7_19_G  sd_RGB_10_7_19_B  
0                237         57.046799         57.085175         50.170302  
1                240         53.171706         54.352414         48.476502  
2                207         62.698556         64.904915         50.130367  
3                210         49.951415         54.026298         41.406264  
4                207         49.409935         49.958817         42.085108  

[5 rows x 23 columns]

In [6]:
# rename columns to be compatible with pipeline
data.columns.values[0] = 'ID'
data.columns.values[1] = 'Class'
data.head()

ID      Class  Min_RGB_10_7_19_R  Min_RGB_10_7_19_G  \
0  EL19_5166  15758.250                  0                  0   
1  EL19_6459  15053.580                  0                  0   
2  EL19_5715  15064.320                 19                 17   
3  EL19_6527  13713.625                 27                 28   
4  EL19_5847  17317.575                 35                 27   

   Min_RGB_10_7_19_B  1stQu_RGB_10_7_19_R  1stQu_RGB_10_7_19_G  \
0                  0                64.75                56.75   
1                  0                70.00                61.00   
2                 17                74.25                72.50   
3                 25                99.75                99.25   
4                 24                75.25                66.25   

   1stQu_RGB_10_7_19_B  Median_RGB_10_7_19_R  Median_RGB_10_7_19_G  ...  \
0                52.00                  93.0                  85.0  ...   
1                54.50                  96.0                  91.0  ...   
2                55.25                 120.5                 122.0  ...   
3                75.75                 131.0                 127.5  ...   
4                60.25                 100.5                  92.0  ...   

   Mean_RGB_10_7_19_B  3rdQu_RGB_10_7_19_R  3rdQu_RGB_10_7_19_G  \
0           87.992647               138.00               129.25   
1           91.296296               134.50               127.00   
2           96.047619               175.75               180.00   
3          107.442857               177.00               182.00   
4           92.301587               152.00               144.00   

   3rdQu_RGB_10_7_19_B  Max_RGB_10_7_19_R  Max_RGB_10_7_19_G  \
0               113.50                254                254   
1               117.50                251                255   
2               131.00                250                251   
3               137.50                240                246   
4               124.75                235                232   

   Max_RGB_10_7_19_B  sd_RGB_10_7_19_R  sd_RGB_10_7_19_G  sd_RGB_10_7_19_B  
0                237         57.046799         57.085175         50.170302  
1                240         53.171706         54.352414         48.476502  
2                207         62.698556         64.904915         50.130367  
3                210         49.951415         54.026298         41.406264  
4                207         49.409935         49.958817         42.085108  

[5 rows x 23 columns]

In [61]:
# write out new df 
data.to_csv('module2_data/featureTable.txt',index=False,sep='\t')

In [62]:
# define a testing set
!python ~/GitHub/SL-ML-pipeline/test_set.py -df module2_data/featureTable.txt -type r -p 0.1 -save module2_data/test_instances.txt

Holding out 10.0 percent
117 instances in test set
finished!


In [65]:
# train the model
!python ~/GitHub/SL-ML-pipeline/ML_regression.py -df module2_data/featureTable.txt -alg RF -y_name Class -test module2_data/test_instances.txt  

Model built using 1168 instances
Removing test instances to apply model on later...
Snapshot of data being used:
                   Y  Min_RGB_10_7_19_R  ...  Min_RGB_10_7_19_B  1stQu_RGB_10_7_19_R
ID                                       ...                                        
EL19_5166  15758.250                  0  ...                  0                64.75
EL19_6459  15053.580                  0  ...                  0                70.00
EL19_5715  15064.320                 19  ...                 17                74.25
EL19_6527  13713.625                 27  ...                 25                99.75
EL19_5847  17317.575                 35  ...                 24                75.25

[5 rows x 5 columns]


===>  Grid search started  <===
Round 1 of 10
^C
Traceback (most recent call last):
  File "/Users/SLotreck/GitHub/SL-ML-pipeline/ML_regression.py", line 486, in <module>
    main()
  File "/Users/SLotreck/GitHub/SL-ML-pipeline/ML_regression.py", line 257, in main
   

**Model training took too long on my own machine, ran as a job in hpcc:** <br>
    
    #!/bin/bash --login
    ########## SBATCH Lines for Resource Request ##########

    #SBATCH --time=03:59:00             # limit of wall clock time - how long the job will run (same as -t)
    #SBATCH --nodes=1                 # number of different nodes - could be an exact number or a range of nodes (same as -N)
    #SBATCH --cpus-per-task=2           # number of CPUs (or cores) per task (same as -c)
    #SBATCH --mem-per-cpu=2G            # memory required per allocated CPU (or core) - amount of memory (in bytes)
    #SBATCH --job-name droneML      # you can give your job a name for easier identification (same as -J)

    ########## Command Lines to Run ##########
    cd ~/Classes/module-2-LUA                  ### change to the directory where your code is located

    srun python ~/Shiu_lab/Pipelines/ML-Pipeline/ML_regression.py -df ~/Classes/module-2-LUA/module2_data/featureTable.txt -alg RF -y_name Class -test ~/Classes/module-2-LUA/module2_data/test_instances.txt           ### call your executable (similar to mpirun)

    scontrol show job $SLURM_JOB_ID     ### write job information to SLURM output file
    js -j $SLURM_JOB_ID                 ### write resource usage to SLURM output file (powetools command)   


**Performance metrics can be found in RESULTS_reg.txt**

# Approach 2: RGB from Sept 2

In [285]:
data2 = pd.read_csv('module2_data/CleanData_InputML_9_2_19.txt',sep = '\t')
data2.head()

PlotID  Accession Genotype AltGenotype  EarLeafArea  TotalLeafAreaPlot  \
0  EL19_5166  Ames10254     GE54        GE54      630.330          15758.250   
1  EL19_6459  Ames10254     GE54        GE54      557.540          15053.580   
2  EL19_5715  Ames10256    GE129       GE129      470.760          15064.320   
3  EL19_6527  Ames10256    GE129       GE129      442.375          13713.625   
4  EL19_5847  Ames12726     NC13        NC13      524.775          17317.575   

   Min_RGB_9_2_19_R  Min_RGB_9_2_19_G  Min_RGB_9_2_19_B  1stQu_RGB_9_2_19_R  \
0                31                30                17               70.50   
1                38                42                27               78.50   
2                24                27                24               49.25   
3                31                31                21               75.75   
4                43                42                24              108.00   

   ...  Mean_RGB_9_2_19_B  3rdQu_RGB_9_2_19_R  3rdQu_RGB_9_2_19_G  \
0  ...          88.352941              148.25              168.25   
1  ...          91.748148              159.50              173.50   
2  ...          76.325397              116.00              138.50   
3  ...          70.400000              130.25              156.00   
4  ...         106.253968              172.75              197.00   

   3rdQu_RGB_9_2_19_B  Max_RGB_9_2_19_R  Max_RGB_9_2_19_G  Max_RGB_9_2_19_B  \
0              125.25               212               241               193   
1              124.00               231               248               210   
2              100.75               210               238               194   
3               86.00               235               252               190   
4              142.75               230               252               206   

   sd_RGB_9_2_19_R  sd_RGB_9_2_19_G  sd_RGB_9_2_19_B  
0        47.504573        54.802771        45.263595  
1        48.251739        53.523626        44.651840  
2        46.832961        52.173168        37.776782  
3        42.543650        51.038173        33.286375  
4        43.680928        48.946763        41.876759  

[5 rows x 27 columns]

In [286]:
# use PlotID as ID, TotalLeafAreaPlot as label, drop all other non-data columns 
data2.drop(labels = ['Accession','Genotype','AltGenotype','EarLeafArea'],axis = 1,inplace=True)
data2.head()

PlotID  TotalLeafAreaPlot  Min_RGB_9_2_19_R  Min_RGB_9_2_19_G  \
0  EL19_5166          15758.250                31                30   
1  EL19_6459          15053.580                38                42   
2  EL19_5715          15064.320                24                27   
3  EL19_6527          13713.625                31                31   
4  EL19_5847          17317.575                43                42   

   Min_RGB_9_2_19_B  1stQu_RGB_9_2_19_R  1stQu_RGB_9_2_19_G  \
0                17               70.50               78.00   
1                27               78.50               92.00   
2                24               49.25               65.00   
3                21               75.75               83.75   
4                24              108.00              112.00   

   1stQu_RGB_9_2_19_B  Median_RGB_9_2_19_R  Median_RGB_9_2_19_G  ...  \
0               50.00                109.0                120.0  ...   
1               52.50                113.0                119.0  ...   
2               44.50                 77.0                 90.5  ...   
3               47.75                 99.5                118.0  ...   
4               74.25                134.0                137.0  ...   

   Mean_RGB_9_2_19_B  3rdQu_RGB_9_2_19_R  3rdQu_RGB_9_2_19_G  \
0          88.352941              148.25              168.25   
1          91.748148              159.50              173.50   
2          76.325397              116.00              138.50   
3          70.400000              130.25              156.00   
4         106.253968              172.75              197.00   

   3rdQu_RGB_9_2_19_B  Max_RGB_9_2_19_R  Max_RGB_9_2_19_G  Max_RGB_9_2_19_B  \
0              125.25               212               241               193   
1              124.00               231               248               210   
2              100.75               210               238               194   
3               86.00               235               252               190   
4              142.75               230               252               206   

   sd_RGB_9_2_19_R  sd_RGB_9_2_19_G  sd_RGB_9_2_19_B  
0        47.504573        54.802771        45.263595  
1        48.251739        53.523626        44.651840  
2        46.832961        52.173168        37.776782  
3        42.543650        51.038173        33.286375  
4        43.680928        48.946763        41.876759  

[5 rows x 23 columns]

In [5]:
# rename columns to be compatible with pipeline
# data2.columns.values[0] = 'ID'
# data2.columns.values[1] = 'Class'
# data2.head()

ID      Class  Min_RGB_9_2_19_R  Min_RGB_9_2_19_G  Min_RGB_9_2_19_B  \
0  EL19_5166  15758.250                31                30                17   
1  EL19_6459  15053.580                38                42                27   
2  EL19_5715  15064.320                24                27                24   
3  EL19_6527  13713.625                31                31                21   
4  EL19_5847  17317.575                43                42                24   

   1stQu_RGB_9_2_19_R  1stQu_RGB_9_2_19_G  1stQu_RGB_9_2_19_B  \
0               70.50               78.00               50.00   
1               78.50               92.00               52.50   
2               49.25               65.00               44.50   
3               75.75               83.75               47.75   
4              108.00              112.00               74.25   

   Median_RGB_9_2_19_R  Median_RGB_9_2_19_G  ...  Mean_RGB_9_2_19_B  \
0                109.0                120.0  ...          88.352941   
1                113.0                119.0  ...          91.748148   
2                 77.0                 90.5  ...          76.325397   
3                 99.5                118.0  ...          70.400000   
4                134.0                137.0  ...         106.253968   

   3rdQu_RGB_9_2_19_R  3rdQu_RGB_9_2_19_G  3rdQu_RGB_9_2_19_B  \
0              148.25              168.25              125.25   
1              159.50              173.50              124.00   
2              116.00              138.50              100.75   
3              130.25              156.00               86.00   
4              172.75              197.00              142.75   

   Max_RGB_9_2_19_R  Max_RGB_9_2_19_G  Max_RGB_9_2_19_B  sd_RGB_9_2_19_R  \
0               212               241               193        47.504573   
1               231               248               210        48.251739   
2               210               238               194        46.832961   
3               235               252               190        42.543650   
4               230               252               206        43.680928   

   sd_RGB_9_2_19_G  sd_RGB_9_2_19_B  
0        54.802771        45.263595  
1        53.523626        44.651840  
2        52.173168        37.776782  
3        51.038173        33.286375  
4        48.946763        41.876759  

[5 rows x 23 columns]

In [6]:
# write out new df 
data2.to_csv('module2_data/featureTable2.txt',index=False,sep='\t')

Will define testing set and train on hpcc

Define testing set: <br>

    python ~/Shiu_lab/Pipelines/ML-Pipeline/test_set.py -df module2_data/featureTable2.txt -type r -p 0.1 -save module2_data/test_instances2.txt

Run training script: 

    #!/bin/bash --login
    ########## SBATCH Lines for Resource Request ##########

    #SBATCH --time=03:59:00             # limit of wall clock time - how long the job will run (same as -t)
    #SBATCH --nodes=1                 # number of different nodes - could be an exact number or a range of nodes (same as -N)
    #SBATCH --cpus-per-task=2           # number of CPUs (or cores) per task (same as -c)
    #SBATCH --mem-per-cpu=2G            # memory required per allocated CPU (or core) - amount of memory (in bytes)
    #SBATCH --job-name droneML2      # you can give your job a name for easier identification (same as -J)

    ########## Command Lines to Run ##########
    cd ~/Classes/module-2-LUA                  ### change to the directory where your code is located

    srun python ~/Shiu_lab/Pipelines/ML-Pipeline/ML_regression.py -df ~/Classes/module-2-LUA/module2_data/featureTable2.txt -alg RF -y_name Class -test ~/Classes/module-2-LUA/module2_data/test_instances2.txt           ### call your executable (similar to mpirun)

    scontrol show job $SLURM_JOB_ID     ### write job information to SLURM output file
    js -j $SLURM_JOB_ID                 ### write resource usage to SLURM output file (powetools command) 

# Approach 3: RGB, multispectral, and height from Sept 2 and Jul 2, with composite features

Composite features: the difference between the same feature on Sept 7 and Jul 2

In [268]:
# import and clean Jul2 data
dataJul = pd.read_csv('module2_data/CleanData_InputML_7_2_19.txt',sep='\t')
dataJul.drop(labels = ['Accession','Genotype','AltGenotype','EarLeafArea'],axis = 1,inplace=True)
dataJul.head()

PlotID  TotalLeafAreaPlot  Min_RGB_7_2_19_R  Min_RGB_7_2_19_G  \
0  EL19_5166          15758.250                70                77   
1  EL19_6459          15053.580                 0                 0   
2  EL19_5715          15064.320                50                67   
3  EL19_6527          13713.625                66                79   
4  EL19_5847          17317.575                30                48   

   Min_RGB_7_2_19_B  1stQu_RGB_7_2_19_R  1stQu_RGB_7_2_19_G  \
0                74              129.75              127.75   
1                 0              117.50              120.00   
2                41              120.00              118.25   
3                67              130.00              128.00   
4                20              101.00              117.50   

   1stQu_RGB_7_2_19_B  Median_RGB_7_2_19_R  Median_RGB_7_2_19_G  ...  \
0              123.75                156.0                149.5  ...   
1              112.00                146.0                140.0  ...   
2              115.00                140.0                134.0  ...   
3              123.00                145.0                138.5  ...   
4               97.25                144.0                139.0  ...   

   Mean_RGB_7_2_19_B  3rdQu_RGB_7_2_19_R  3rdQu_RGB_7_2_19_G  \
0         142.816176              174.25              171.00   
1         126.733333              157.50              150.50   
2         125.095238              149.00              145.75   
3         131.464286              158.00              152.00   
4         119.507937              159.00              150.00   

   3rdQu_RGB_7_2_19_B  Max_RGB_7_2_19_R  Max_RGB_7_2_19_G  Max_RGB_7_2_19_B  \
0              166.25               199               195               192   
1              146.00               185               176               171   
2              141.75               168               169               164   
3              148.00               184               176               175   
4              144.00               185               175               170   

   sd_RGB_7_2_19_R  sd_RGB_7_2_19_G  sd_RGB_7_2_19_B  
0        30.044321        27.084854        28.427999  
1        34.656559        29.685533        30.312354  
2        28.688913        20.628759        25.477026  
3        27.769992        20.018772        24.497074  
4        37.649111        28.088361        33.736330  

[5 rows x 23 columns]

In [287]:
dataSept = data2
dataSept.head()

PlotID  TotalLeafAreaPlot  Min_RGB_9_2_19_R  Min_RGB_9_2_19_G  \
0  EL19_5166          15758.250                31                30   
1  EL19_6459          15053.580                38                42   
2  EL19_5715          15064.320                24                27   
3  EL19_6527          13713.625                31                31   
4  EL19_5847          17317.575                43                42   

   Min_RGB_9_2_19_B  1stQu_RGB_9_2_19_R  1stQu_RGB_9_2_19_G  \
0                17               70.50               78.00   
1                27               78.50               92.00   
2                24               49.25               65.00   
3                21               75.75               83.75   
4                24              108.00              112.00   

   1stQu_RGB_9_2_19_B  Median_RGB_9_2_19_R  Median_RGB_9_2_19_G  ...  \
0               50.00                109.0                120.0  ...   
1               52.50                113.0                119.0  ...   
2               44.50                 77.0                 90.5  ...   
3               47.75                 99.5                118.0  ...   
4               74.25                134.0                137.0  ...   

   Mean_RGB_9_2_19_B  3rdQu_RGB_9_2_19_R  3rdQu_RGB_9_2_19_G  \
0          88.352941              148.25              168.25   
1          91.748148              159.50              173.50   
2          76.325397              116.00              138.50   
3          70.400000              130.25              156.00   
4         106.253968              172.75              197.00   

   3rdQu_RGB_9_2_19_B  Max_RGB_9_2_19_R  Max_RGB_9_2_19_G  Max_RGB_9_2_19_B  \
0              125.25               212               241               193   
1              124.00               231               248               210   
2              100.75               210               238               194   
3               86.00               235               252               190   
4              142.75               230               252               206   

   sd_RGB_9_2_19_R  sd_RGB_9_2_19_G  sd_RGB_9_2_19_B  
0        47.504573        54.802771        45.263595  
1        48.251739        53.523626        44.651840  
2        46.832961        52.173168        37.776782  
3        42.543650        51.038173        33.286375  
4        43.680928        48.946763        41.876759  

[5 rows x 23 columns]

In [180]:
# add in multispectral data 
multispecMeans = pd.read_csv('module2_data/multispec_means_obs2019.csv')
multispecMeans.head()

ZONE_CODE             id  BLUE_7_2_19  GREEN_7_2_19  NIR_7_2_19  \
0          1  EL19_fill-1-1          NaN           NaN         NaN   
1          2  EL19_fill-1-2     0.076648      0.137853    0.342161   
2          3  EL19_fill-1-3     0.071476      0.119751    0.335271   
3          4  EL19_fill-1-4     0.068427      0.115552    0.339573   
4          5  EL19_fill-1-5     0.063336      0.109313    0.340780   

   RED_7_2_19  REDEDGE_7_2_19  BLUE_7_15_19  GREEN_7_15_19  NIR_7_15_19  ...  \
0         NaN             NaN           NaN            NaN          NaN  ...   
1    0.151015        0.239997      0.052558       0.110591     0.387516  ...   
2    0.139666        0.214942      0.056186       0.109478     0.427869  ...   
3    0.129405        0.208753      0.053380       0.103086     0.421574  ...   
4    0.117529        0.201306      0.044824       0.095132     0.420984  ...   

   COUNT_7_15_19  AREA_7_15_19  COUNT_7_28_19  AREA_7_28_19  COUNT_8_12_19  \
0            NaN           NaN          182.0      3.727446          797.0   
1            3.0      0.059812          322.0      6.594712         1223.0   
2           65.0      1.295934          493.0     10.096873         1299.0   
3          154.0      3.070366          713.0     14.602577         1393.0   
4          437.0      8.712661         1218.0     24.945215         1618.0   

   AREA_8_12_19  COUNT_9_2_19  AREA_9_2_19  COUNT_9_11_19  AREA_9_11_19  
0     15.860894          1695    31.925016         1491.0     30.263872  
1     24.338611          1834    34.543055         1673.0     33.958054  
2     25.851068          1701    32.038025         1596.0     32.395131  
3     27.721738          1695    31.925016         1549.0     31.441139  
4     32.199405          1802    33.940341         1666.0     33.815970  

[5 rows x 44 columns]

In [181]:
# clean out non-experimental measurements
excludeFill = []
IDs = multispecMeans.id
for name in IDs:
    if 'fill' in name:
        excludeFill.append(name)

multispecMeansExp = multispecMeans[~multispecMeans['id'].isin(excludeFill)]
multispecMeansExp.head()

ZONE_CODE              id  BLUE_7_2_19  GREEN_7_2_19  NIR_7_2_19  \
18         19  EL19_5098-1-19     0.072238      0.115038    0.312846   
19         20  EL19_5099-1-20     0.067315      0.117966    0.338393   
41         42  EL19_5139-2-41     0.063710      0.107563    0.316869   
42         43  EL19_5138-2-40     0.063918      0.110355    0.330111   
43         44  EL19_5137-2-39     0.066221      0.111373    0.336597   

    RED_7_2_19  REDEDGE_7_2_19  BLUE_7_15_19  GREEN_7_15_19  NIR_7_15_19  ...  \
18    0.138869        0.201090      0.044831       0.086719     0.334554  ...   
19    0.123294        0.213633      0.041522       0.089272     0.370372  ...   
41    0.119378        0.194726      0.040626       0.080731     0.342263  ...   
42    0.119961        0.202151      0.039323       0.079479     0.339438  ...   
43    0.125830        0.202628      0.038287       0.075498     0.362253  ...   

    COUNT_7_15_19  AREA_7_15_19  COUNT_7_28_19  AREA_7_28_19  COUNT_8_12_19  \
18          220.0      4.386237         1202.0     24.617527         1614.0   
19          872.0     17.385448         1446.0     29.614763         1728.0   
41          665.0     13.258398         1358.0     27.812481         1448.0   
42          683.0     13.617272         1148.0     23.511582         1321.0   
43          902.0     17.983571         1505.0     30.823111         1692.0   

    AREA_8_12_19  COUNT_9_2_19  AREA_9_2_19  COUNT_9_11_19  AREA_9_11_19  
18     32.119802          1796    33.827332         1677.0     34.039244  
19     34.388487          1798    33.865002         1702.0     34.546687  
41     28.816279          1771    33.356462         1510.0     30.649528  
42     26.288884          1380    25.992048           42.0      0.852503  
43     33.672060          1713    32.264043          642.0     13.031124  

[5 rows x 44 columns]

In [182]:
from collections import defaultdict

In [183]:
# clean plot id's to match RGB
replacements = defaultdict(str)
for name in multispecMeansExp.id:
    splitStr = name.split('-')
    replacements[name] = splitStr[0]
    
multispecMeansExp.replace(replacements,inplace=True)
multispecMeansExp.head()

/anaconda3/envs/ML_pipeline/lib/python3.7/site-packages/pandas/core/frame.py:4263: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  method=method,


ZONE_CODE         id  BLUE_7_2_19  GREEN_7_2_19  NIR_7_2_19  RED_7_2_19  \
18         19  EL19_5098     0.072238      0.115038    0.312846    0.138869   
19         20  EL19_5099     0.067315      0.117966    0.338393    0.123294   
41         42  EL19_5139     0.063710      0.107563    0.316869    0.119378   
42         43  EL19_5138     0.063918      0.110355    0.330111    0.119961   
43         44  EL19_5137     0.066221      0.111373    0.336597    0.125830   

    REDEDGE_7_2_19  BLUE_7_15_19  GREEN_7_15_19  NIR_7_15_19  ...  \
18        0.201090      0.044831       0.086719     0.334554  ...   
19        0.213633      0.041522       0.089272     0.370372  ...   
41        0.194726      0.040626       0.080731     0.342263  ...   
42        0.202151      0.039323       0.079479     0.339438  ...   
43        0.202628      0.038287       0.075498     0.362253  ...   

    COUNT_7_15_19  AREA_7_15_19  COUNT_7_28_19  AREA_7_28_19  COUNT_8_12_19  \
18          220.0      4.386237         1202.0     24.617527         1614.0   
19          872.0     17.385448         1446.0     29.614763         1728.0   
41          665.0     13.258398         1358.0     27.812481         1448.0   
42          683.0     13.617272         1148.0     23.511582         1321.0   
43          902.0     17.983571         1505.0     30.823111         1692.0   

    AREA_8_12_19  COUNT_9_2_19  AREA_9_2_19  COUNT_9_11_19  AREA_9_11_19  
18     32.119802          1796    33.827332         1677.0     34.039244  
19     34.388487          1798    33.865002         1702.0     34.546687  
41     28.816279          1771    33.356462         1510.0     30.649528  
42     26.288884          1380    25.992048           42.0      0.852503  
43     33.672060          1713    32.264043          642.0     13.031124  

[5 rows x 44 columns]

In [184]:
# select dates 
dates = []
for col in multispecMeansExp.columns.values:
    if col != 'id' and '7_2_19' not in col:
        if '9_2_19' not in col:
            dates.append(col)
multispecMeansExp.drop(dates,axis=1,inplace=True)
multispecMeansExp.head()

/anaconda3/envs/ML_pipeline/lib/python3.7/site-packages/pandas/core/frame.py:4102: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


id  BLUE_7_2_19  GREEN_7_2_19  NIR_7_2_19  RED_7_2_19  \
18  EL19_5098     0.072238      0.115038    0.312846    0.138869   
19  EL19_5099     0.067315      0.117966    0.338393    0.123294   
41  EL19_5139     0.063710      0.107563    0.316869    0.119378   
42  EL19_5138     0.063918      0.110355    0.330111    0.119961   
43  EL19_5137     0.066221      0.111373    0.336597    0.125830   

    REDEDGE_7_2_19  BLUE_9_2_19  GREEN_9_2_19  NIR_9_2_19  RED_9_2_19  \
18        0.201090     0.038164      0.086298    0.496407    0.070752   
19        0.213633     0.029780      0.073491    0.453850    0.049416   
41        0.194726     0.028718      0.071867    0.389276    0.057498   
42        0.202151     0.036708      0.079013    0.368206    0.084656   
43        0.202628     0.035979      0.084395    0.453303    0.090688   

    REDEDGE_9_2_19  COUNT_7_2_19  AREA_7_2_19  COUNT_9_2_19  AREA_9_2_19  
18        0.202326         135.0     2.752898          1796    33.827332  
19        0.162900         431.0     8.788883          1798    33.865002  
41        0.161120         419.0     8.544181          1771    33.356462  
42        0.178422         406.0     8.279087          1380    25.992048  
43        0.231845         447.0     9.115152          1713    32.264043

In [185]:
multispecMeansExp.drop(['COUNT_7_2_19','AREA_7_2_19','COUNT_9_2_19','AREA_9_2_19'],axis=1,inplace=True)
multispecMeansExp.head()

id  BLUE_7_2_19  GREEN_7_2_19  NIR_7_2_19  RED_7_2_19  \
18  EL19_5098     0.072238      0.115038    0.312846    0.138869   
19  EL19_5099     0.067315      0.117966    0.338393    0.123294   
41  EL19_5139     0.063710      0.107563    0.316869    0.119378   
42  EL19_5138     0.063918      0.110355    0.330111    0.119961   
43  EL19_5137     0.066221      0.111373    0.336597    0.125830   

    REDEDGE_7_2_19  BLUE_9_2_19  GREEN_9_2_19  NIR_9_2_19  RED_9_2_19  \
18        0.201090     0.038164      0.086298    0.496407    0.070752   
19        0.213633     0.029780      0.073491    0.453850    0.049416   
41        0.194726     0.028718      0.071867    0.389276    0.057498   
42        0.202151     0.036708      0.079013    0.368206    0.084656   
43        0.202628     0.035979      0.084395    0.453303    0.090688   

    REDEDGE_9_2_19  
18        0.202326  
19        0.162900  
41        0.161120  
42        0.178422  
43        0.231845

In [270]:
# separate into jul and sept dfs 
sept = []
jul = []
for col in multispecMeansExp.columns.values:
    if col != 'id' and '7_2_19' in col:
        jul.append(col)
    elif col != 'id' and '9_2_19' in col:
        sept.append(col)
        
septMultispec = multispecMeansExp.drop(jul,axis=1)
julMultispec = multispecMeansExp.drop(sept,axis=1)

In [187]:
julMultispec.head()

id  BLUE_7_2_19  GREEN_7_2_19  NIR_7_2_19  RED_7_2_19  \
18  EL19_5098     0.072238      0.115038    0.312846    0.138869   
19  EL19_5099     0.067315      0.117966    0.338393    0.123294   
41  EL19_5139     0.063710      0.107563    0.316869    0.119378   
42  EL19_5138     0.063918      0.110355    0.330111    0.119961   
43  EL19_5137     0.066221      0.111373    0.336597    0.125830   

    REDEDGE_7_2_19  
18        0.201090  
19        0.213633  
41        0.194726  
42        0.202151  
43        0.202628

In [188]:
septMultispec.head()

id  BLUE_9_2_19  GREEN_9_2_19  NIR_9_2_19  RED_9_2_19  \
18  EL19_5098     0.038164      0.086298    0.496407    0.070752   
19  EL19_5099     0.029780      0.073491    0.453850    0.049416   
41  EL19_5139     0.028718      0.071867    0.389276    0.057498   
42  EL19_5138     0.036708      0.079013    0.368206    0.084656   
43  EL19_5137     0.035979      0.084395    0.453303    0.090688   

    REDEDGE_9_2_19  
18        0.202326  
19        0.162900  
41        0.161120  
42        0.178422  
43        0.231845

## Note! Do not do this:
    julMultispec.columns.values[0] = 'PlotID'
This only superficially renames the column, will throw lots of errors in the long run <br>
Do this instead:

In [277]:
# combine RGB and multispec for each date 
septMultispec.rename(columns = {'id':'PlotID'},inplace=True) # it was all because of this.......
julMultispec.rename(columns = {'id':'PlotID'},inplace=True)

In [190]:
septMultispec.head()

PlotID  BLUE_9_2_19  GREEN_9_2_19  NIR_9_2_19  RED_9_2_19  \
18  EL19_5098     0.038164      0.086298    0.496407    0.070752   
19  EL19_5099     0.029780      0.073491    0.453850    0.049416   
41  EL19_5139     0.028718      0.071867    0.389276    0.057498   
42  EL19_5138     0.036708      0.079013    0.368206    0.084656   
43  EL19_5137     0.035979      0.084395    0.453303    0.090688   

    REDEDGE_9_2_19  
18        0.202326  
19        0.162900  
41        0.161120  
42        0.178422  
43        0.231845

In [288]:
dataSept.head()

PlotID  TotalLeafAreaPlot  Min_RGB_9_2_19_R  Min_RGB_9_2_19_G  \
0  EL19_5166          15758.250                31                30   
1  EL19_6459          15053.580                38                42   
2  EL19_5715          15064.320                24                27   
3  EL19_6527          13713.625                31                31   
4  EL19_5847          17317.575                43                42   

   Min_RGB_9_2_19_B  1stQu_RGB_9_2_19_R  1stQu_RGB_9_2_19_G  \
0                17               70.50               78.00   
1                27               78.50               92.00   
2                24               49.25               65.00   
3                21               75.75               83.75   
4                24              108.00              112.00   

   1stQu_RGB_9_2_19_B  Median_RGB_9_2_19_R  Median_RGB_9_2_19_G  ...  \
0               50.00                109.0                120.0  ...   
1               52.50                113.0                119.0  ...   
2               44.50                 77.0                 90.5  ...   
3               47.75                 99.5                118.0  ...   
4               74.25                134.0                137.0  ...   

   Mean_RGB_9_2_19_B  3rdQu_RGB_9_2_19_R  3rdQu_RGB_9_2_19_G  \
0          88.352941              148.25              168.25   
1          91.748148              159.50              173.50   
2          76.325397              116.00              138.50   
3          70.400000              130.25              156.00   
4         106.253968              172.75              197.00   

   3rdQu_RGB_9_2_19_B  Max_RGB_9_2_19_R  Max_RGB_9_2_19_G  Max_RGB_9_2_19_B  \
0              125.25               212               241               193   
1              124.00               231               248               210   
2              100.75               210               238               194   
3               86.00               235               252               190   
4              142.75               230               252               206   

   sd_RGB_9_2_19_R  sd_RGB_9_2_19_G  sd_RGB_9_2_19_B  
0        47.504573        54.802771        45.263595  
1        48.251739        53.523626        44.651840  
2        46.832961        52.173168        37.776782  
3        42.543650        51.038173        33.286375  
4        43.680928        48.946763        41.876759  

[5 rows x 23 columns]

In [278]:
multiID = septMultispec.PlotID
RGBid = dataSept.PlotID 

In [279]:
membership = multiID.isin(RGBid)
membership.value_counts()

True     1168
False     356
Name: PlotID, dtype: int64

According to the above, there should be 356 columns lost when I merge over the PlotID column.

In [289]:
septAll = pd.merge(septMultispec,dataSept,on='PlotID',how='inner')
septAll.head()

PlotID  BLUE_9_2_19  GREEN_9_2_19  NIR_9_2_19  RED_9_2_19  \
0  EL19_5098     0.038164      0.086298    0.496407    0.070752   
1  EL19_5099     0.029780      0.073491    0.453850    0.049416   
2  EL19_5137     0.035979      0.084395    0.453303    0.090688   
3  EL19_5136     0.036586      0.093272    0.442665    0.071858   
4  EL19_5135     0.035176      0.087523    0.476621    0.067009   

   REDEDGE_9_2_19  TotalLeafAreaPlot  Min_RGB_9_2_19_R  Min_RGB_9_2_19_G  \
0        0.202326           9484.425                45                42   
1        0.162900          12293.775                18                20   
2        0.231845          14374.080                57                50   
3        0.198094          17118.975                 0                 0   
4        0.183911          12882.870                26                26   

   Min_RGB_9_2_19_B  ...  Mean_RGB_9_2_19_B  3rdQu_RGB_9_2_19_R  \
0                18  ...          92.792857              146.50   
1                 9  ...          75.264286              122.25   
2                33  ...          81.650794              148.50   
3                 0  ...          83.230159              152.25   
4                12  ...          80.484127              138.00   

   3rdQu_RGB_9_2_19_G  3rdQu_RGB_9_2_19_B  Max_RGB_9_2_19_R  Max_RGB_9_2_19_G  \
0               156.5              120.00               232               249   
1               147.0              107.75               208               234   
2               139.5               96.75               214               200   
3               163.0              121.00               253               252   
4               155.0              112.75               206               224   

   Max_RGB_9_2_19_B  sd_RGB_9_2_19_R  sd_RGB_9_2_19_G  sd_RGB_9_2_19_B  
0               208        42.533084        46.597850        39.447351  
1               186        48.645087        55.489441        42.471423  
2               161        31.379674        32.661026        27.496128  
3               211        51.863063        57.827926        44.949601  
4               186        43.354802        49.163288        39.318339  

[5 rows x 28 columns]

In [201]:
# do it for july!
julMultispec.head()

PlotID  BLUE_7_2_19  GREEN_7_2_19  NIR_7_2_19  RED_7_2_19  \
18  EL19_5098     0.072238      0.115038    0.312846    0.138869   
19  EL19_5099     0.067315      0.117966    0.338393    0.123294   
41  EL19_5139     0.063710      0.107563    0.316869    0.119378   
42  EL19_5138     0.063918      0.110355    0.330111    0.119961   
43  EL19_5137     0.066221      0.111373    0.336597    0.125830   

    REDEDGE_7_2_19  
18        0.201090  
19        0.213633  
41        0.194726  
42        0.202151  
43        0.202628

In [202]:
dataJul.head()

PlotID  TotalLeafAreaPlot  Min_RGB_7_2_19_R  Min_RGB_7_2_19_G  \
0  EL19_5166          15758.250                70                77   
1  EL19_6459          15053.580                 0                 0   
2  EL19_5715          15064.320                50                67   
3  EL19_6527          13713.625                66                79   
4  EL19_5847          17317.575                30                48   

   Min_RGB_7_2_19_B  1stQu_RGB_7_2_19_R  1stQu_RGB_7_2_19_G  \
0                74              129.75              127.75   
1                 0              117.50              120.00   
2                41              120.00              118.25   
3                67              130.00              128.00   
4                20              101.00              117.50   

   1stQu_RGB_7_2_19_B  Median_RGB_7_2_19_R  Median_RGB_7_2_19_G  ...  \
0              123.75                156.0                149.5  ...   
1              112.00                146.0                140.0  ...   
2              115.00                140.0                134.0  ...   
3              123.00                145.0                138.5  ...   
4               97.25                144.0                139.0  ...   

   Mean_RGB_7_2_19_B  3rdQu_RGB_7_2_19_R  3rdQu_RGB_7_2_19_G  \
0         142.816176              174.25              171.00   
1         126.733333              157.50              150.50   
2         125.095238              149.00              145.75   
3         131.464286              158.00              152.00   
4         119.507937              159.00              150.00   

   3rdQu_RGB_7_2_19_B  Max_RGB_7_2_19_R  Max_RGB_7_2_19_G  Max_RGB_7_2_19_B  \
0              166.25               199               195               192   
1              146.00               185               176               171   
2              141.75               168               169               164   
3              148.00               184               176               175   
4              144.00               185               175               170   

   sd_RGB_7_2_19_R  sd_RGB_7_2_19_G  sd_RGB_7_2_19_B  
0        30.044321        27.084854        28.427999  
1        34.656559        29.685533        30.312354  
2        28.688913        20.628759        25.477026  
3        27.769992        20.018772        24.497074  
4        37.649111        28.088361        33.736330  

[5 rows x 23 columns]

In [280]:
julAll = pd.merge(julMultispec,dataJul,on='PlotID',how='inner')
julAll.head()

PlotID  BLUE_7_2_19  GREEN_7_2_19  NIR_7_2_19  RED_7_2_19  \
0  EL19_5098     0.072238      0.115038    0.312846    0.138869   
1  EL19_5099     0.067315      0.117966    0.338393    0.123294   
2  EL19_5137     0.066221      0.111373    0.336597    0.125830   
3  EL19_5136     0.063849      0.112825    0.338212    0.121923   
4  EL19_5135     0.064075      0.114575    0.337937    0.120117   

   REDEDGE_7_2_19  TotalLeafAreaPlot  Min_RGB_7_2_19_R  Min_RGB_7_2_19_G  \
0        0.201090           9484.425                49                36   
1        0.213633          12293.775                82                91   
2        0.202628          14374.080                44                64   
3        0.205385          17118.975                 0                 0   
4        0.202816          12882.870                50                68   

   Min_RGB_7_2_19_B  ...  Mean_RGB_7_2_19_B  3rdQu_RGB_7_2_19_R  \
0                35  ...         149.192857              175.00   
1                80  ...         143.607143              171.00   
2                38  ...         132.222222              178.75   
3                 0  ...         128.460317              169.75   
4                47  ...         130.301587              169.00   

   3rdQu_RGB_7_2_19_G  3rdQu_RGB_7_2_19_B  Max_RGB_7_2_19_R  Max_RGB_7_2_19_G  \
0              170.25              167.00               199               191   
1              170.00              164.00               193               188   
2              168.00              158.75               199               192   
3              162.75              151.00               194               186   
4              160.00              150.00               194               185   

   Max_RGB_7_2_19_B  sd_RGB_7_2_19_R  sd_RGB_7_2_19_G  sd_RGB_7_2_19_B  
0               189        21.914676        22.081763        22.513592  
1               184        25.506519        21.074686        24.476886  
2               181        37.709945        31.736326        33.947463  
3               177        36.905880        30.568593        33.295922  
4               175        32.804738        25.265756        29.169373  

[5 rows x 28 columns]

In [299]:
# pair up column names for subtraction 
septCols = pd.Series(septAll.columns.values) 
julCols = pd.Series(julAll.columns.values)
subCols = pd.concat([septCols,julCols],axis=1)
subCols.drop([0,6],axis=0,inplace=True)
subCols

0                    1
1           BLUE_9_2_19          BLUE_7_2_19
2          GREEN_9_2_19         GREEN_7_2_19
3            NIR_9_2_19           NIR_7_2_19
4            RED_9_2_19           RED_7_2_19
5        REDEDGE_9_2_19       REDEDGE_7_2_19
7      Min_RGB_9_2_19_R     Min_RGB_7_2_19_R
8      Min_RGB_9_2_19_G     Min_RGB_7_2_19_G
9      Min_RGB_9_2_19_B     Min_RGB_7_2_19_B
10   1stQu_RGB_9_2_19_R   1stQu_RGB_7_2_19_R
11   1stQu_RGB_9_2_19_G   1stQu_RGB_7_2_19_G
12   1stQu_RGB_9_2_19_B   1stQu_RGB_7_2_19_B
13  Median_RGB_9_2_19_R  Median_RGB_7_2_19_R
14  Median_RGB_9_2_19_G  Median_RGB_7_2_19_G
15  Median_RGB_9_2_19_B  Median_RGB_7_2_19_B
16    Mean_RGB_9_2_19_R    Mean_RGB_7_2_19_R
17    Mean_RGB_9_2_19_G    Mean_RGB_7_2_19_G
18    Mean_RGB_9_2_19_B    Mean_RGB_7_2_19_B
19   3rdQu_RGB_9_2_19_R   3rdQu_RGB_7_2_19_R
20   3rdQu_RGB_9_2_19_G   3rdQu_RGB_7_2_19_G
21   3rdQu_RGB_9_2_19_B   3rdQu_RGB_7_2_19_B
22     Max_RGB_9_2_19_R     Max_RGB_7_2_19_R
23     Max_RGB_9_2_19_G     Max_RGB_7_2_19_G
24     Max_RGB_9_2_19_B     Max_RGB_7_2_19_B
25      sd_RGB_9_2_19_R      sd_RGB_7_2_19_R
26      sd_RGB_9_2_19_G      sd_RGB_7_2_19_G
27      sd_RGB_9_2_19_B      sd_RGB_7_2_19_B

In [321]:
# create full feature table 
AllFeatures = pd.merge(julAll,septAll,on='PlotID',how='inner')
AllFeatures.head()

PlotID  BLUE_7_2_19  GREEN_7_2_19  NIR_7_2_19  RED_7_2_19  \
0  EL19_5098     0.072238      0.115038    0.312846    0.138869   
1  EL19_5099     0.067315      0.117966    0.338393    0.123294   
2  EL19_5137     0.066221      0.111373    0.336597    0.125830   
3  EL19_5136     0.063849      0.112825    0.338212    0.121923   
4  EL19_5135     0.064075      0.114575    0.337937    0.120117   

   REDEDGE_7_2_19  TotalLeafAreaPlot_x  Min_RGB_7_2_19_R  Min_RGB_7_2_19_G  \
0        0.201090             9484.425                49                36   
1        0.213633            12293.775                82                91   
2        0.202628            14374.080                44                64   
3        0.205385            17118.975                 0                 0   
4        0.202816            12882.870                50                68   

   Min_RGB_7_2_19_B  ...  Mean_RGB_9_2_19_B  3rdQu_RGB_9_2_19_R  \
0                35  ...          92.792857              146.50   
1                80  ...          75.264286              122.25   
2                38  ...          81.650794              148.50   
3                 0  ...          83.230159              152.25   
4                47  ...          80.484127              138.00   

   3rdQu_RGB_9_2_19_G  3rdQu_RGB_9_2_19_B  Max_RGB_9_2_19_R  Max_RGB_9_2_19_G  \
0               156.5              120.00               232               249   
1               147.0              107.75               208               234   
2               139.5               96.75               214               200   
3               163.0              121.00               253               252   
4               155.0              112.75               206               224   

   Max_RGB_9_2_19_B  sd_RGB_9_2_19_R  sd_RGB_9_2_19_G  sd_RGB_9_2_19_B  
0               208        42.533084        46.597850        39.447351  
1               186        48.645087        55.489441        42.471423  
2               161        31.379674        32.661026        27.496128  
3               211        51.863063        57.827926        44.949601  
4               186        43.354802        49.163288        39.318339  

[5 rows x 55 columns]

In [322]:
# perform subtraction and add columns onto the end of comined df 
for i in subCols.itertuples(): 
    sept = i[1]
    jul = i[2]
    featureName = 'diff_'+sept+'_'+jul
    newCol = septAll[sept] - julAll[jul]
    newCol.name = featureName
    AllFeatures = pd.concat([AllFeatures, newCol], axis=1)
    

In [323]:
AllFeatures.head()

PlotID  BLUE_7_2_19  GREEN_7_2_19  NIR_7_2_19  RED_7_2_19  \
0  EL19_5098     0.072238      0.115038    0.312846    0.138869   
1  EL19_5099     0.067315      0.117966    0.338393    0.123294   
2  EL19_5137     0.066221      0.111373    0.336597    0.125830   
3  EL19_5136     0.063849      0.112825    0.338212    0.121923   
4  EL19_5135     0.064075      0.114575    0.337937    0.120117   

   REDEDGE_7_2_19  TotalLeafAreaPlot_x  Min_RGB_7_2_19_R  Min_RGB_7_2_19_G  \
0        0.201090             9484.425                49                36   
1        0.213633            12293.775                82                91   
2        0.202628            14374.080                44                64   
3        0.205385            17118.975                 0                 0   
4        0.202816            12882.870                50                68   

   Min_RGB_7_2_19_B  ...  diff_Mean_RGB_9_2_19_B_Mean_RGB_7_2_19_B  \
0                35  ...                                -56.400000   
1                80  ...                                -68.342857   
2                38  ...                                -50.571429   
3                 0  ...                                -45.230159   
4                47  ...                                -49.817460   

   diff_3rdQu_RGB_9_2_19_R_3rdQu_RGB_7_2_19_R  \
0                                      -28.50   
1                                      -48.75   
2                                      -30.25   
3                                      -17.50   
4                                      -31.00   

   diff_3rdQu_RGB_9_2_19_G_3rdQu_RGB_7_2_19_G  \
0                                      -13.75   
1                                      -23.00   
2                                      -28.50   
3                                        0.25   
4                                       -5.00   

   diff_3rdQu_RGB_9_2_19_B_3rdQu_RGB_7_2_19_B  \
0                                      -47.00   
1                                      -56.25   
2                                      -62.00   
3                                      -30.00   
4                                      -37.25   

   diff_Max_RGB_9_2_19_R_Max_RGB_7_2_19_R  \
0                                      33   
1                                      15   
2                                      15   
3                                      59   
4                                      12   

   diff_Max_RGB_9_2_19_G_Max_RGB_7_2_19_G  \
0                                      58   
1                                      46   
2                                       8   
3                                      66   
4                                      39   

   diff_Max_RGB_9_2_19_B_Max_RGB_7_2_19_B  \
0                                      19   
1                                       2   
2                                     -20   
3                                      34   
4                                      11   

   diff_sd_RGB_9_2_19_R_sd_RGB_7_2_19_R  diff_sd_RGB_9_2_19_G_sd_RGB_7_2_19_G  \
0                             20.618408                             24.516087   
1                             23.138568                             34.414755   
2                             -6.330271                              0.924699   
3                             14.957182                             27.259334   
4                             10.550064                             23.897532   

   diff_sd_RGB_9_2_19_B_sd_RGB_7_2_19_B  
0                             16.933759  
1                             17.994537  
2                             -6.451334  
3                             11.653679  
4                             10.148966  

[5 rows x 81 columns]

In [324]:
# manually verify one of the differences 
blueJulZero = AllFeatures.iloc[0,1]
blueSeptZero = AllFeatures.iloc[0,28]
blueDiff = AllFeatures.iloc[0,55]
blueDiff == blueSeptZero - blueJulZero

True

In [325]:
# check that the two TLA columns match and then get rid of one 
x = AllFeatures.TotalLeafAreaPlot_x
y = AllFeatures.TotalLeafAreaPlot_y
x.equals(y)

True

In [326]:
AllFeatures.drop('TotalLeafAreaPlot_x',axis=1,inplace=True)

In [327]:
AllFeatures.rename(columns = {'TotalLeafAreaPlot_y':'TotalLeafAreaPlot'},inplace=True)

In [328]:
# rearrange so TLA is the second column 
featCols = AllFeatures.columns.values.tolist()
print(len(featCols))
groundIndex = featCols.index('TotalLeafAreaPlot')
groundTruth = featCols[groundIndex]
featCols.pop(groundIndex)
print(len(featCols))
featCols.insert(1, groundTruth)
featCols

80
79


['PlotID',
 'TotalLeafAreaPlot',
 'BLUE_7_2_19',
 'GREEN_7_2_19',
 'NIR_7_2_19',
 'RED_7_2_19',
 'REDEDGE_7_2_19',
 'Min_RGB_7_2_19_R',
 'Min_RGB_7_2_19_G',
 'Min_RGB_7_2_19_B',
 '1stQu_RGB_7_2_19_R',
 '1stQu_RGB_7_2_19_G',
 '1stQu_RGB_7_2_19_B',
 'Median_RGB_7_2_19_R',
 'Median_RGB_7_2_19_G',
 'Median_RGB_7_2_19_B',
 'Mean_RGB_7_2_19_R',
 'Mean_RGB_7_2_19_G',
 'Mean_RGB_7_2_19_B',
 '3rdQu_RGB_7_2_19_R',
 '3rdQu_RGB_7_2_19_G',
 '3rdQu_RGB_7_2_19_B',
 'Max_RGB_7_2_19_R',
 'Max_RGB_7_2_19_G',
 'Max_RGB_7_2_19_B',
 'sd_RGB_7_2_19_R',
 'sd_RGB_7_2_19_G',
 'sd_RGB_7_2_19_B',
 'BLUE_9_2_19',
 'GREEN_9_2_19',
 'NIR_9_2_19',
 'RED_9_2_19',
 'REDEDGE_9_2_19',
 'Min_RGB_9_2_19_R',
 'Min_RGB_9_2_19_G',
 'Min_RGB_9_2_19_B',
 '1stQu_RGB_9_2_19_R',
 '1stQu_RGB_9_2_19_G',
 '1stQu_RGB_9_2_19_B',
 'Median_RGB_9_2_19_R',
 'Median_RGB_9_2_19_G',
 'Median_RGB_9_2_19_B',
 'Mean_RGB_9_2_19_R',
 'Mean_RGB_9_2_19_G',
 'Mean_RGB_9_2_19_B',
 '3rdQu_RGB_9_2_19_R',
 '3rdQu_RGB_9_2_19_G',
 '3rdQu_RGB_9_2_19_B',
 '

In [329]:
AllFeatures = AllFeatures[featCols]
AllFeatures.head()

PlotID  TotalLeafAreaPlot  BLUE_7_2_19  GREEN_7_2_19  NIR_7_2_19  \
0  EL19_5098           9484.425     0.072238      0.115038    0.312846   
1  EL19_5099          12293.775     0.067315      0.117966    0.338393   
2  EL19_5137          14374.080     0.066221      0.111373    0.336597   
3  EL19_5136          17118.975     0.063849      0.112825    0.338212   
4  EL19_5135          12882.870     0.064075      0.114575    0.337937   

   RED_7_2_19  REDEDGE_7_2_19  Min_RGB_7_2_19_R  Min_RGB_7_2_19_G  \
0    0.138869        0.201090                49                36   
1    0.123294        0.213633                82                91   
2    0.125830        0.202628                44                64   
3    0.121923        0.205385                 0                 0   
4    0.120117        0.202816                50                68   

   Min_RGB_7_2_19_B  ...  diff_Mean_RGB_9_2_19_B_Mean_RGB_7_2_19_B  \
0                35  ...                                -56.400000   
1                80  ...                                -68.342857   
2                38  ...                                -50.571429   
3                 0  ...                                -45.230159   
4                47  ...                                -49.817460   

   diff_3rdQu_RGB_9_2_19_R_3rdQu_RGB_7_2_19_R  \
0                                      -28.50   
1                                      -48.75   
2                                      -30.25   
3                                      -17.50   
4                                      -31.00   

   diff_3rdQu_RGB_9_2_19_G_3rdQu_RGB_7_2_19_G  \
0                                      -13.75   
1                                      -23.00   
2                                      -28.50   
3                                        0.25   
4                                       -5.00   

   diff_3rdQu_RGB_9_2_19_B_3rdQu_RGB_7_2_19_B  \
0                                      -47.00   
1                                      -56.25   
2                                      -62.00   
3                                      -30.00   
4                                      -37.25   

   diff_Max_RGB_9_2_19_R_Max_RGB_7_2_19_R  \
0                                      33   
1                                      15   
2                                      15   
3                                      59   
4                                      12   

   diff_Max_RGB_9_2_19_G_Max_RGB_7_2_19_G  \
0                                      58   
1                                      46   
2                                       8   
3                                      66   
4                                      39   

   diff_Max_RGB_9_2_19_B_Max_RGB_7_2_19_B  \
0                                      19   
1                                       2   
2                                     -20   
3                                      34   
4                                      11   

   diff_sd_RGB_9_2_19_R_sd_RGB_7_2_19_R  diff_sd_RGB_9_2_19_G_sd_RGB_7_2_19_G  \
0                             20.618408                             24.516087   
1                             23.138568                             34.414755   
2                             -6.330271                              0.924699   
3                             14.957182                             27.259334   
4                             10.550064                             23.897532   

   diff_sd_RGB_9_2_19_B_sd_RGB_7_2_19_B  
0                             16.933759  
1                             17.994537  
2                             -6.451334  
3                             11.653679  
4                             10.148966  

[5 rows x 80 columns]

In [330]:
# rename columns to be compatible with pipeline
AllFeatures.rename(columns = {'PlotID':'ID','TotalLeafAreaPlot':'Class'},inplace=True)
AllFeatures.head()

ID      Class  BLUE_7_2_19  GREEN_7_2_19  NIR_7_2_19  RED_7_2_19  \
0  EL19_5098   9484.425     0.072238      0.115038    0.312846    0.138869   
1  EL19_5099  12293.775     0.067315      0.117966    0.338393    0.123294   
2  EL19_5137  14374.080     0.066221      0.111373    0.336597    0.125830   
3  EL19_5136  17118.975     0.063849      0.112825    0.338212    0.121923   
4  EL19_5135  12882.870     0.064075      0.114575    0.337937    0.120117   

   REDEDGE_7_2_19  Min_RGB_7_2_19_R  Min_RGB_7_2_19_G  Min_RGB_7_2_19_B  ...  \
0        0.201090                49                36                35  ...   
1        0.213633                82                91                80  ...   
2        0.202628                44                64                38  ...   
3        0.205385                 0                 0                 0  ...   
4        0.202816                50                68                47  ...   

   diff_Mean_RGB_9_2_19_B_Mean_RGB_7_2_19_B  \
0                                -56.400000   
1                                -68.342857   
2                                -50.571429   
3                                -45.230159   
4                                -49.817460   

   diff_3rdQu_RGB_9_2_19_R_3rdQu_RGB_7_2_19_R  \
0                                      -28.50   
1                                      -48.75   
2                                      -30.25   
3                                      -17.50   
4                                      -31.00   

   diff_3rdQu_RGB_9_2_19_G_3rdQu_RGB_7_2_19_G  \
0                                      -13.75   
1                                      -23.00   
2                                      -28.50   
3                                        0.25   
4                                       -5.00   

   diff_3rdQu_RGB_9_2_19_B_3rdQu_RGB_7_2_19_B  \
0                                      -47.00   
1                                      -56.25   
2                                      -62.00   
3                                      -30.00   
4                                      -37.25   

   diff_Max_RGB_9_2_19_R_Max_RGB_7_2_19_R  \
0                                      33   
1                                      15   
2                                      15   
3                                      59   
4                                      12   

   diff_Max_RGB_9_2_19_G_Max_RGB_7_2_19_G  \
0                                      58   
1                                      46   
2                                       8   
3                                      66   
4                                      39   

   diff_Max_RGB_9_2_19_B_Max_RGB_7_2_19_B  \
0                                      19   
1                                       2   
2                                     -20   
3                                      34   
4                                      11   

   diff_sd_RGB_9_2_19_R_sd_RGB_7_2_19_R  diff_sd_RGB_9_2_19_G_sd_RGB_7_2_19_G  \
0                             20.618408                             24.516087   
1                             23.138568                             34.414755   
2                             -6.330271                              0.924699   
3                             14.957182                             27.259334   
4                             10.550064                             23.897532   

   diff_sd_RGB_9_2_19_B_sd_RGB_7_2_19_B  
0                             16.933759  
1                             17.994537  
2                             -6.451334  
3                             11.653679  
4                             10.148966  

[5 rows x 80 columns]

In [331]:
# write out as csv
AllFeatures.to_csv('module2_data/allButHeightFeatures.csv',index=False)

This csv was then passed to getHeights.py on hpcc to add height data, and the model was trained with the commmands and job script below:

    python ~/Shiu_lab/Pipelines/ML-Pipeline/test_set.py -df module2_data/FinalFeatureTableSeptJulDiff.csv -type r -p 0.1 -sep ',' -save module2_data/test_instances_SeptJulDiff.txt

    #!/bin/bash --login
    ########## SBATCH Lines for Resource Request ##########

    #SBATCH --time=03:59:00             # limit of wall clock time - how long the job will run (same as -t)
    #SBATCH --nodes=1                 # number of different nodes - could be an exact number or a range of nodes (same as -N)
    #SBATCH --cpus-per-task=2           # number of CPUs (or cores) per task (same as -c)
    #SBATCH --mem-per-cpu=2G            # memory required per allocated CPU (or core) - amount of memory (in bytes)
    #SBATCH --job-name droneML3      # you can give your job a name for easier identification (same as -J)

    ########## Command Lines to Run ##########
    cd ~/Classes/module-2-LUA                  ### change to the directory where your code is located

    srun python ~/Shiu_lab/Pipelines/ML-Pipeline/ML_regression.py -df ~/Classes/module-2-LUA/module2_data/FinalFeatureTableSeptJulDiff.csv -alg RF -y_name Class -sep ',' -test ~/Classes/module-2-LUA/module2_data/test_instances_SeptJulDiff.txt           ### call your executable (similar to mpirun)

    scontrol show job $SLURM_JOB_ID     ### write job information to SLURM output file
        js -j $SLURM_JOB_ID                 ### write resource usage to SLURM output file (powetools command) 